Code/Engine/Foundation/Basics.h

```diff
@@ -10,14 +10,19 @@
 #include <Foundation/UserConfig.h>
 
 // Configure the DLL Import/Export Define
+// EZ_FOUNDATION_DLL_CPP17 should be used if a class or struct has additional c++17 attributes e.g.
+// struct [[nodiscard]] EZ_FOUNDATION_DLL_CPP17 StructName { ... };
 #if EZ_ENABLED(EZ_COMPILE_ENGINE_AS_DLL)
 #  ifdef BUILDSYSTEM_BUILDING_FOUNDATION_LIB
 #    define EZ_FOUNDATION_DLL EZ_DECL_EXPORT
+#    define EZ_FOUNDATION_DLL_CPP17 EZ_DECL_EXPORT_CPP17
```

Comment:
Looking for a better name for this. Maybe `EZ_FOUNDATION_DLL_CPP_ATTR` ?
And no we can't always use the [[attr]] syntax as this is not valid in all places we need.

Reply:
So when can you use the one Syntax and when the other?  Functions versus classes? Maybe that can help to find a better name.

Reply:
Looking at it closer it seems the only problems are statements like this:
`friend EZ_FOUNDATION_DLL void ezRun(ezApplication* pApplicationInstance);`
of which we have 6 in foundation.
I have the hunch that in this case the `EZ_FOUNDATION_DLL` does not actually do anything, as its only a friend declaration. Let me try that out

Reply:
Ok so the main problem was that the [[gnu::visibility("default")]] attribute would conflict with `__attribute__((alignas(X)))`.
I pretty much removed all usage of our `EZ_ALIGN` macro and replaced it with c++11 `alignas`.

We can now use `[[gnu::visibility("default")]]` everywhere.

Reply:
The only reason EZ_ALIGN still exists is because we use EZ_ALIGN_16 in our shader source code.
We could replace EZ_ALIGN_16 with something like EZ_SHADER_STRUCT ?

Reply:
Sounds good.

Reply:
I was able to get completely rid of most EZ_ALIGN and related macros (only EZ_ALIGNMENT_OF and EZ_MINIMUM_ALIGNMENT still exist).

Unfortunately declaring a friend function on windows is a special case and we now have a new macro for that. In foundation for example we now have EZ_FOUNDATION_DLL_FRIEND. In core EZ_CORE_DLL_FRIEND. We only have like 8 cases where we need these macros, so most developers should not need to care about them.